core/src/main/java/org/infinispan/distribution/LocalizedCacheTopology.java

```diff
@@ -0,0 +1,190 @@
+package org.infinispan.distribution;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.infinispan.commons.hash.MurmurHash3;
+import org.infinispan.commons.util.Immutables;
+import org.infinispan.commons.util.IntSet;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.distribution.ch.ConsistentHash;
+import org.infinispan.distribution.ch.KeyPartitioner;
+import org.infinispan.distribution.ch.impl.ReplicatedConsistentHash;
+import org.infinispan.remoting.transport.Address;
+import org.infinispan.topology.CacheTopology;
+
+/**
+ * Extends {@link CacheTopology} with information about keys owned by the local node.
+ *
+ * @author Dan Berindei
+ * @since 9.0
+ */
+public class LocalizedCacheTopology extends CacheTopology {
```

Comment:
That plan's perfectly fine; the cache mode does not change, so the switch should be in the user of the routing information. `LocalizedCacheTopology` should provide only information "who's read/write" owner, an that should be the line. Otherwise, you will end up with a second `ClusteringDependentLogic`, just using if-else instead of polymorphism.